### DISCUSSION TITLE
Load functions from external js files in a quarto doc

### DISCUSSION THREAD

**Discussion by boshek** [quarto-dev/quarto-cli#6570] [Q&A] [answered]
### Description

This is probably such a noob question but...

How can we load functions in an external file so that they are available at render-time for a quarto doc? 

Say I have this `functions.js` file:
```js
sayGoodbye = function(x) {
    return 1 - x;
  }
 ```
 
 And then I have this quarto doc:
 ````
 ---
format: 
  html:
    self-contained: true
---

## Working function
```{ojs}
sayHello = function(x) {
  return 1 +  x;
}

sayHello(3);
sayHello(13);
```

## Attempts to load it
```{ojs}
import { sayGoodbye } from './functions.js';

sayGoodbye(3);
```

```{ojs}
import { sayGoodbye } from 'functions.js';

sayGoodbye(3);
```

```{ojs}
require('functions.js');

sayGoodbye(3);
```
````

These all failed at the `import`/`require` stages. Any suggestions? It is so nice both from a code modularity perspective and also from a readability to have those tucked away in a separate file. I feel like i must be missing something obvious.

**Comment by mcanouil** [comment]
To note this is unrelated to Quarto.
You should take a look on https://observablehq.com/ and related resources.

This being said, here is an example:

<table>
<tr><th>Quarto document</th><th>JavaScript</th></tr>
<tr>
<td>

````qmd
---
title: "Quarto Playground"
format: html
---

This is a playground for Quarto.

```{ojs}
import { myFunction } from "./myModule.js";

myFunction();
```

````

</td>
<td>

````js
function myFunction() {
  // function body
  return 1;
}

export { myFunction };

````

</td>
</tr>
</table>

Edit: from the documentation <https://quarto.org/docs/interactive/ojs/code-reuse.html#modules>

**Reply by boshek** [reply]
Thanks for this. I do, however, thinkit  is related to Quarto. I believe that Quarto users who are also taking advantage of `ojs` chunks will definitely find the ability to store code and function in local js files useful.

**Reply by mcanouil** [reply]
If a user wants to use Observable.js it's not on Quarto to document the features of that tool I am afraid.
Everything I showed is pure Observable.js code/syntax.
The same applies for Python, Julia, and R.
It's on the user to look for the language documentation.

**Reply by boshek** [reply]
Sorry to stretch this but AKAIK, Quarto represents one of (the?) the only locally hosted means of using Observable. To do what I was suggesting using the Observable platform, you'd just create another Observable notebook. But (and of course I could definitely be wrong) but using Observable this way is unique to Quarto (though obviously not to JavaScript). So I do still think that it has a unique quarto element to it.  

> it's not on Quarto to document the features of that tool

Have I misinterpreted this space? My question was definitely not a request for you folks to document this but more a forum-style question. Apologies if this is the wrong space to do that.

**Reply by mcanouil** [reply]
> Have I misinterpreted this space? My question was definitely not a request for you folks to document this but more a forum-style question.

Not what I meant or you would not have had an answer.
Importing file or module is an Observable thing not a Quarto thing thus the best place is platform dedicated to that tool even if without being expert we might be able to help here.
It’s exactly the same as asking R/Python questions.

> Apologies if this is the wrong space to do that.

no reason to apologise.
I am really just pointing out that the question and solution is only about Observable and that in general, it’s better to ask on the dedicated platforms, because it’s on those platforms that the users of Observable are.

To note, I basically found / double checked the answer by searching the web on how to import file module in Observable.

**Reply by cscheid** [reply]
@mcanouil Sorry, but @boshek is right here; our support for OJS goes beyond what ObservableHQ does. We explain a bit of that https://quarto.org/docs/interactive/ojs/

(@boshek, sorry about this; Quarto is a big system, and this part is not very widely advertised.)

**Reply by mcanouil** [reply]
@cscheid so the import for JavaScript is not an Observable feature?
No needs to be sorry, as I said, I am no Observable expert so I double checked and found the same thing without any link to Quarto, so I just assumed it was Observable feature.

I have to admit that now I am a bit confused about what Quarto adds or not to Observable ...


Should a new section in <https://quarto.org/docs/interactive/ojs/libraries.html> be added to show the example I shared then?

**Reply by cscheid** [reply, marked as answer]
@boshek one issue here is that OJS doesn't work well with`self-contained: true`, exactly because of javascript imports.

You'll find that the following works:

<img width="715" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/285675/ca778d7a-59eb-46c2-a004-33d3d07a953f">

````
---
title: ojs import test
---


```{ojs}
import {myFunction} from "./myModule.js";

myFunction(3);
```
````

myModule.js:

````
export function myFunction(x)
{
    return x * x;
}
````

However, you _will_ need to serve this page from a http or https server (instead of the local file:// scheme), because javascript module imports are only supported in the http and https protocols.

**Reply by cscheid** [reply]
> @cscheid so the import for JavaScript is not an Observable feature?

It's not _only_ an observable feature. We enrich it by allowing users to import directly from .js files, from .ts files, and from other .qmd files, in addition to importing it from other observablehq notebooks.

**Reply by mcanouil** [reply]
With what you explained about importing local files.
I definitely think this should be part of <https://quarto.org/docs/interactive/ojs/libraries.html>.

**Reply by cscheid** [reply]
We do document it here: https://quarto.org/docs/interactive/ojs/code-reuse.html

**Reply by mcanouil** [reply]
So I completely missed this from the documentation even in search as the actual answer to the main question is <https://quarto.org/docs/interactive/ojs/code-reuse.html#modules>.

**Reply by boshek** [reply]
Thanks for all this follow-up - classic fail I made me reprex too simple. I am re-posting because I expect that the biggest use here is folks that want to re-use `Plot` code. However I can't quite see how I can import the library in the separate file. Basically I am looking for the `dplyr::` equivalent from R. Here is the revised reprex:

````
---
format: html
---

```{r}
ojs_define(ojs_pen = palmerpenguins::penguins)
```

```{ojs}
pen = transpose(ojs_pen)
import { plotterHelper } from "./functions.js";

plotterHelper(pen)
```
````

and then **functions.js**
```js
// I guess this doesn't work because I'm not using Observable yet?
import { Plot } from '@observablehq/plot';

function plotterHelper(data) {
  return Plot.plot({
    grid: true,
    inset: 10,
    color: {legend: true},
    marks: [
      Plot.frame(),
      Plot.dot(data, {x: "bill_length_mm", y: "bill_depth_mm", stroke: "green"})
    ]
  });
}

export { plotterHelper };
```

I get this error:
> Failed to resolve module specifier "@observablehq/plot". Relative references must start with either "/", "./", or "../".

Also tried:
`import * as Plot from '@observablehq/plot';` & `Plot = await import("https://esm.sh/@observablehq/plot");`

**Reply by cscheid** [reply]
Your JS files need to be plain javascript. You're importing from a `.js` file, not an `.ojs` file.

If you want reuse OJS code, then you need to import from a separate notebook, via the .qmd syntax.

**Reply by cscheid** [reply]
The following is _OJS_ syntax, and doesn't support `import * as Plot`:

```
import { Plot } from '@observablehq/plot';
```

In a plain .js file, you'll need to import from esm.sh, which _does_ support `import *`. Still, it might be the case that you can't use it, because that depends on how the export is done by their library. You don't need top-level await, a straightforward import statement works.

```
import { Plot } from "https://esm.sh/@observablehq/plot"; (or `* as Plot`, I don't remember how they expose their exports)
```

or to send along the Plot global from OJS:

```js
function plotterHelper(data, Plot) {
  return Plot.plot({
    grid: true,
    inset: 10,
    color: {legend: true},
    marks: [
      Plot.frame(),
      Plot.dot(data, {x: "bill_length_mm", y: "bill_depth_mm", stroke: "green"})
    ]
  });
}
```

And from OJS you'd call it like `plotterHelper(data, Plot)`. There's nothing much we can do about this impedance mismatch, unfortunately. You have to be mindful of how Javascript features work, which parts of your code are OJS, and which are JS, and treat them appropriately.

**Reply by boshek** [reply]
> If you want reuse OJS code, then you need to import from a separate notebook, via the .qmd syntax.

This makes _perfect_ sense. I get it now. And then that is analogous to importing a notebook from observable.

**Reply by cscheid** [reply]
> And then that is analogous to importing a notebook from observable.

That's exactly right, yes!
